### PR TITLE
ElasticsearchGraphMaker first implementation

### PR DESCRIPTION
**Pull request by ethan-rambacher**

### CODE REVIEW

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                throw new AlreadyExistsException("Graph '" + lowerName + "' already exists");
+            }
+        } else {
+            // there is no graph with this name yet
+            CreateIndexRequest request = new CreateIndexRequest(lowerName);
+            request.source(elasticsearchIndexSettings, XContentType.JSON);
+            try {
+                client.indices().create(request, RequestOptions.DEFAULT);
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java]: Use SLF4J before you get into this habit..

Comment (gordom6) [reply src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java]: In this case you'd want to bubble up the exception anyway. Wrap it in a RuntimeException maybe.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/elasticsearchIndexSettings.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/elasticsearchIndexSettings.json:1]: Don't need this copy?

File: src/test/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphTest.java [hunk 3]
```diff
@@ -1,21 +1,43 @@
...
         super(name);
     }
 
-    public static TestSuite suite()
-    { return new TestSuite( ElasticsearchGraphTest.class ); }
+    public static TestSuite suite() {
+        return new TestSuite(ElasticsearchGraphTest.class);
+    }
 
     public Graph getGraph() {
```
Comment (gordom6) [inline src/test/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphTest.java:22]: You seem to be missing @Override annotations. IntelliJ will add these on save with the "Save Actions" plugin.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java [hunk 4]
```diff
@@ -1,16 +1,180 @@
...
+        jsonMap.put("subject", getNodeContent(t.getSubject()));
+        jsonMap.put("predicate", getNodeContent(t.getPredicate()));
+        jsonMap.put("object", getNodeContent(t.getObject()));
 
+        try {
+            IndexRequest request = new IndexRequest(this.name).source(jsonMap);
+            this.client.index(request, RequestOptions.DEFAULT);
+        } catch (IOException e) {
+            logger.error("Error indexing triple: {}", t, e);
+            e.printStackTrace();
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java]: Don't need this too.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java [hunk 5]
```diff
@@ -1,16 +1,180 @@
...
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.query(this.constructTripleMatchingQuery(triple));
+
+        SearchRequest searchRequest = new SearchRequest();
+        searchRequest.indices(this.name);
+        searchRequest.source(searchSourceBuilder);
+
+        try {
+            SearchResponse searchResponse = this.client.search(searchRequest, RequestOptions.DEFAULT);
+            List<Triple> triple_results = new ArrayList<>();
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java]: Camel case..

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java [hunk 6]
```diff
@@ -1,16 +1,180 @@
...
+            for (SearchHit hit : searchResponse.getHits().getHits()) {
+                Map<String, Object> fields = hit.getSourceAsMap();
+                String s = (String) fields.get("subject");
+                String p = (String) fields.get("predicate");
+                String o = (String) fields.get("object");
+
+                triple_results.add(Triple.create(createNode(s), createNode(p), createNode(o)));
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java]: Subjects and predicates can't be literals. Predicates can't be blank nodes. These functions should reflect that.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java [hunk 7]
```diff
@@ -1,16 +1,180 @@
...
+     * subject, predicate, or object can be Node.ANY, meaning that the given field will match anything
+     *
+     * @param triple the triple that the query must match
+     * @return a QueryBuilder that will match that triple
+     */
+    private QueryBuilder constructTripleMatchingQuery(Triple triple) {
+        BoolQueryBuilder queryBuilder = new BoolQueryBuilder();
+
+        if (!triple.getSubject().equals(Node.ANY)) {
+            MatchQueryBuilder matchSubjectQueryBuilder = new MatchQueryBuilder("subject", getNodeContent(triple.getSubject()));
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java]: You want an (exact) term query, not a (full-text) match query.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java [hunk 8]
```diff
@@ -1,16 +1,180 @@
...
+            searchSourceBuilder.query(QueryBuilders.matchAllQuery());
+            countRequest.source(searchSourceBuilder);
+
+            CountResponse response = this.client.count(countRequest, RequestOptions.DEFAULT);
+            return (int) response.getCount();
+        } catch (IOException e) {
+            logger.error("Could not retrieve size of graph '{}'", this.name, e);
+        }
+        // TODO: Is this an appropriate default return value??
+        return -1;
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java]: No, rethrow the IOException as a  RuntimeException. throw new RuntimeException(e).

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java [hunk 9]
```diff
@@ -1,16 +1,180 @@
...
 
-    public ExtendedIterator<Triple> find() {
-        throw new UnsupportedOperationException();
+    /**
+     * Returns a blank, literal, or URI node for the given string
+     *
+     * @param s the value of the node
+     * @return a node corresponding to s
+     */
+    private Node createNode(String s) {
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java]: static

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java [hunk 10]
```diff
@@ -1,115 +1,237 @@
...
 import java.util.Set;
+import java.util.stream.Collectors;
 
 public class ElasticsearchGraphMaker extends BaseGraphMaker {
-    public Set<String> graphNames = new LinkedHashSet<>();
-    public RestHighLevelClient client;
-    public String elasticsearchIndexSettings;
-    public final String settingsPath = "elasticsearchIndexSettings.json";
+    private Logger logger = LoggerFactory.getLogger(ElasticsearchGraphMaker.class);
+    private Set<String> graphNames;
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java:35]: All of these should be final. Use local variables in the constructor if you need to, then assign to the member.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java [hunk 11]
```diff
@@ -1,115 +1,237 @@
...
+        this.client = new RestHighLevelClient(RestClient.builder(httpHosts));
 
         // Get all the graphs currently on this Elasticsearch cluster
         // and add them to the locally tracked set
         GetIndexRequest request = new GetIndexRequest("_all");
-        GetIndexResponse response = client.indices().get(request, RequestOptions.DEFAULT);
-        graphNames.addAll(Arrays.asList(response.getIndices()));
+
+        try {
+            GetIndexResponse response = this.client.indices().get(request, RequestOptions.DEFAULT);
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java:68]: This try-catch block to a separate method.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java [hunk 12]
```diff
@@ -1,115 +1,237 @@
...
+     *               instead of returning the existing graph
      * @return
      */
+    @Override
     public Graph createGraph(String name, boolean strict) throws AlreadyExistsException {
-        // Elasticsearch only allows lowercase index names
-        String lowerName = name.toLowerCase();
+        // Encode the given graph name into a name safe for Elasticsearch
+        String validIndexName = encodeIndexName(name);
+        if (validIndexName.startsWith("_")) {
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java]: Still have this?

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java [hunk 13]
```diff
@@ -1,115 +1,237 @@
...
-     * @param strict if true, throw AlreadyExistsException a graph if a graph with the given name already exists
+     * @param strict if true, throw AlreadyExistsException a graph if a graph with the given name does not exist,
+     *               instead of creating the graph
      * @return
      */
+    @Override
     public Graph openGraph(String name, boolean strict) {
+        // Elasticsearch only allows lowercase index names
+        String validIndexName = encodeIndexName(name);
+        if (validIndexName.startsWith("_")) {
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java]: Repeated code.

File: src/test/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMakerTest.java [hunk 14]
```diff
@@ -4,30 +4,36 @@
...
+                    new HttpHost("localhost", 9200, "http")
             );
 
             // Create and return the graph
-            return graphMaker;
+            return this.gf;
         } catch (Exception e) {
             // Error reading ES index settings
             e.printStackTrace();
             return null;
```
Comment (gordom6) [inline src/test/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMakerTest.java]: Exception.

**Review by gordom6 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java (hunk 6 above)
Comment (ethan-rambacher) [reply src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java]: If any of those things occur, what should happen? RuntimeException, UnsupportedOperationException, or IllegalArgumentException?

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java [hunk 15]
```diff
@@ -56,11 +62,44 @@ public void performAdd(Triple t) {
...
+        final RefreshRequest refreshRequest = new RefreshRequest(this.name);
+        final SearchRequest searchRequest = new SearchRequest();
+        final SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+
+        searchSourceBuilder.query(this.constructExactTripleMatchingQuery(t));
+        searchRequest.indices(this.name);
+        searchRequest.source(searchSourceBuilder);
+
+        // Try 10 times to refresh Elasticsearch
+        for (int i=0; i<10; ++i) {
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraph.java]: You can factor this loop out into a separate method that takes a java.util.function.Predicate to do the test.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java [hunk 16]
```diff
@@ -143,7 +142,7 @@ public Graph openGraph(String name, boolean strict) {
             // there is no graph with this name yet
             if (strict) {
                 // throw an error instead of creating the graph
-                logger.error("Cannot open graph '{}'; does not exist", validIndexName);
+                this.logger.error("Cannot open graph '{}'; does not exist", validIndexName);
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java:145]: Don't need this prefix.

**Review by gordom6 — COMMENTED** [review]
Looks like it's on the right track.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java (hunk 16 above)
Comment (ethan-rambacher) [reply src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java:145]: Is it bad practice to include the "this" ? I thought it was better to explicitly include it but I'm curious as to what is generally best

Comment (gordom6) [reply src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchGraphMaker.java:145]: It's not bad practice, it's just not common.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchTripleIterator.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.jena.graph.Graph;
+import org.apache.jena.graph.Triple;
+import org.apache.jena.util.iterator.NiceIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+public class ElasticsearchTripleIterator extends NiceIterator<Triple> {
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchTripleIterator.java]: Package private.

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchTripleIterator.java [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    private final Logger logger = LoggerFactory.getLogger(ElasticsearchGraph.class);
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchTripleIterator.java]: static

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchTripleIterator.java [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    private Graph graph;
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchTripleIterator.java]: All final?

File: src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchTripleIterator.java [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+    private Collection<Triple> collection;
+    private Iterator<Triple> iter;
+
+    public static ElasticsearchTripleIterator create(Collection<Triple> c, Graph g) {
+        ElasticsearchTripleIterator iterator = new ElasticsearchTripleIterator();
+        iterator.graph = g;
```
Comment (gordom6) [inline src/main/java/io/github/tetherless_world/jena_elasticsearch/ElasticsearchTripleIterator.java]: Noooooooo

**Review by gordom6 — CHANGES_REQUESTED** [review]

**Review by gordom6 — APPROVED** [review]
Fin.

**Comment by minorg** [comment]
Please merge this so you can start a new branch.

**Pull request merged**